src/core/modules/impl/CacheableStateEvaluator.ts

```diff
@@ -49,28 +49,34 @@ export class CacheableStateEvaluator extends DefaultStateEvaluator {
       throw new Error('Contract tx id not set in the execution context');
     }
 
+    const isFirstEvaluation = cachedState == null;
+    let baseState = isFirstEvaluation ? executionContext.contractDefinition.initState : cachedState.cachedValue.state;
+    const baseValidity = isFirstEvaluation ? {} : cachedState.cachedValue.validity;
+    const baseErrorMessages = isFirstEvaluation ? {} : cachedState.cachedValue.errorMessages;
+
+    if (isFirstEvaluation) {
+      baseState = await executionContext.handler.maybeCallStateConstructor(
+        executionContext.contractDefinition.initState,
+        executionContext
+      );
+    }
+
     if (missingInteractions.length == 0) {
       this.cLogger.info(`No missing interactions ${contractTxId}`);
       if (cachedState) {
-        executionContext.handler?.initState(cachedState.cachedValue.state);
+        executionContext.handler.initState(cachedState.cachedValue.state);
```

Comment:
hmm, with your refactor above - I believe it could be now called like
`executionContext.handler.initState(baseState);` before line 66?

Reply:
or maybe even MUST be called like this, otherwise you might loose changes set in the constructor (line 58)

Reply:
also - why `?` has been removed? handler is not always created and added to executionContext.

Reply:
true true, would fail for rust contracts, which are not implemented for now.
I have removed `?` because `ExecutionContext::handler` is not defined as optional: `handler: Api`

Reply:
Nope it shouldn't called like this because it means if(cachedState) it means that this is not first interaction so baseState === null (constructor won't be called)

Reply:
ok, mea culpa!